platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,927 @@
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <FieldArray
+      name="marriages"
+      render={arrayHelpers => (
+        <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+      )}
+    />
+  )}
+</Page>
+```
+
+For array pages to work we need a couple core things:
+- The path for the page
+- The data for the current item in the array
+
+To get these pieces, `Page` will accept the `path` and `pathParams` props.
+
+`path` will be the URL within the application and must (for array pages) contain
+URL parameters (term pending). URL parameters must be the only token between
+`/`s or at the end of the URL. For example, `/:index/foo` and `/:index` are
+valid path parameters. `/my-:index-item` is an invalid path parameter.
+
+`pathParams` will be an object whose keys map to the path parameters and whose
+values map to the path to the form data. In the above example, the `pathParams`
+will assign `:marriageIndex` in the URL a number based in the current item in
+the `marriages` array.
+
+##### Review page
+To keep Formulate simple, it will not build the review page automatically. It
+will be the responsibility of the engineers to build the review page. This
+decision is subject to change, but as of the time of this writing, I can't think
+of a way to automatically build the review page and meet all of the following
+requirements:
+- The chapters are divided correctly
+- The pages are divided correctly
+- The edit mode for each page is identical to the page in the form flow
+- The review mode for each page displays each field on the page in a data table
+  with the appropriate label and value
+- Each field in the review mode may be overridden when the label or value need
+  to be changed for a better user experience
+- Each page may or may not have a review mode
+
+Instead, Formulate will supply components to aid engineers in rapidly building
+review pages. These components will include:
+
+- `FormReview`
+- `ChapterReview`
+- `PageReview`
+- `FieldReview`
+
+**TODO:** Figure out how submission validation plays in here. How do we run all
+validation on all pages?
+
+#### Save-in-progress
+Pieces of the save-in-progress sub-module are:
+- [`SaveInProgress` component](#component-saveinprogress)
+  - Adds `SiPIndicator` to each page
+  - Provides functions to save and load the form
+- [`SiPIndicator` component](#component-sipindicator)
+  - Allows users to manually save the form
+  - Shows the form save state (saved, saving, error, etc.)
```

Comment:
Can we also include & show the SiP/application ID. So the users can provide a non-PII ID to the help desk allowing engineers an easy method to look up the SiP data.

Reply:
If this is data that's returned by the API, we 100% can show it. :+1: